File: android/app/src/main/java/com/privatepensieve/app/recall/RecallEngine.kt
package com.privatepensieve.app.recall

import com.privatepensieve.app.models.MemoryCard
import com.privatepensieve.app.providers.FakeAIBrain
import java.time.Duration
import java.time.Instant
import kotlin.math.exp
import kotlin.math.max

/**
 * Recall Engine — answers personal memory questions using evidence-bound retrieval.
 * Implements weighted scoring from docs/RECALL_PIPELINE.md.
 * No hallucination. No cloud. Local evidence only.
 */
class RecallEngine {

    enum class QueryClass { MEMORY_RECALL, REFLECTION, GENERAL_FRIEND }

    companion object {
        private const val WEIGHT_LEXICAL = 0.35
        private const val WEIGHT_TAG_MATCH = 0.25
        private const val WEIGHT_RECENCY = 0.15
        private const val WEIGHT_IMPORTANCE = 0.15
        private const val WEIGHT_CONFIDENCE = 0.10
        private const val EVIDENCE_THRESHOLD = 0.15
        const val NO_MEMORY_FALLBACK = "I don't remember you telling me that yet."
    }

    private val aiBrain = FakeAIBrain()

    data class RecallResult(
        val answer: String,
        val evidence: List<MemoryCard>,
        val queryClass: QueryClass
    )

    /**
     * Execute a full recall query.
     * For V1, candidates list is passed in from DAO externally.
     */
    suspend fun recall(question: String, candidates: List<MemoryCard>): RecallResult {
        val queryClass = classifyQuery(question)

        if (queryClass == QueryClass.GENERAL_FRIEND) {
            val reply = aiBrain.generateFriendReply(question, emptyList())
            return RecallResult(reply, emptyList(), queryClass)
        }

        val keywords = extractKeywords(question)
        val scored = candidates
            .map { ScoredCard(it, computeScore(it, keywords)) }
            .sortedByDescending { it.score }
            .filter { it.score >= EVIDENCE_THRESHOLD }

        val topEvidence = scored.take(5).map { it.card }

        if (topEvidence.isEmpty()) {
            return RecallResult(NO_MEMORY_FALLBACK, emptyList(), queryClass)
        }

        val answer = aiBrain.answerFromEvidence(question, topEvidence)
        return RecallResult(answer, topEvidence, queryClass)
    }

    private fun classifyQuery(question: String): QueryClass {
        val lower = question.lowercase()
        val recallKeywords = listOf("what did i", "when did i", "did i say", "did i mention",
            "what was", "who was", "where was", "tell me about", "remind me")
        val reflectionKeywords = listOf("pattern", "how often", "usually", "trend", "frequently")

        if (recallKeywords.any { lower.contains(it) }) return QueryClass.MEMORY_RECALL
        if (reflectionKeywords.any { lower.contains(it) }) return QueryClass.REFLECTION
        return QueryClass.GENERAL_FRIEND
    }

    private fun extractKeywords(question: String): List<String> {
        val stopWords = setOf("the", "a", "an", "is", "was", "are", "were", "i", "my", "me",
            "did", "do", "what", "when", "where", "who", "how", "about", "tell", "remind", "say")
        return question.lowercase().replace(Regex("[^a-z0-9 ]"), "")
            .split(" ").filter { it.length > 2 && it !in stopWords }
    }

    private fun computeScore(card: MemoryCard, keywords: List<String>): Double {
        val lexical = lexicalScore(card, keywords)
        val tagMatch = tagMatchScore(card, keywords)
        val recency = recencyScore(card)
        val importance = card.importanceScore / 10.0
        val confidence = card.confidenceScore

        return (WEIGHT_LEXICAL * lexical) + (WEIGHT_TAG_MATCH * tagMatch) +
               (WEIGHT_RECENCY * recency) + (WEIGHT_IMPORTANCE * importance) +
               (WEIGHT_CONFIDENCE * confidence)
    }

    private fun lexicalScore(card: MemoryCard, keywords: List<String>): Double {
        if (keywords.isEmpty()) return 0.0
        val text = "${card.title} ${card.summary}".lowercase()
        val matches = keywords.count { text.contains(it) }
        return matches.toDouble() / keywords.size
    }

    private fun tagMatchScore(card: MemoryCard, keywords: List<String>): Double {
        if (keywords.isEmpty()) return 0.0
        val allTags = (card.emotionTags + card.topicTags + card.peopleTags + card.goalTags)
            .map { it.lowercase() }
        val matches = keywords.count { kw -> allTags.any { it.contains(kw) } }
        return matches.toDouble() / keywords.size
    }

    private fun recencyScore(card: MemoryCard): Double {
        val daysSince = Duration.between(card.createdAt, Instant.now()).toDays().toDouble()
        return max(0.0, exp(-daysSince / 10.0))
    }

    private data class ScoredCard(val card: MemoryCard, val score: Double)
}

File: ios/PrivatePensieve/Screens/Recall/RecallEngine.swift
// RecallEngine.swift
// Private Pensieve AI — iOS
// Core recall pipeline: query → classify → search → rank → compress → answer
// Implements the weighted scoring formula from docs/RECALL_PIPELINE.md

import Foundation

/// Recall Engine — answers personal memory questions using evidence-bound retrieval.
/// No hallucination. If no evidence passes threshold, returns exact fallback.
final class RecallEngine {

    // MARK: - Query Classification

    enum QueryClass {
        case memoryRecall   // "What did I say about..." — needs factual evidence
        case reflection     // "What patterns do I have..." — needs aggregated evidence
        case generalFriend  // "I'm feeling down" — conversation/support, no assertions
    }

    // MARK: - Scoring Weights (from RECALL_PIPELINE.md)

    private static let weightLexical: Double = 0.35
    private static let weightTagMatch: Double = 0.25
    private static let weightRecency: Double = 0.15
    private static let weightImportance: Double = 0.15
    private static let weightConfidence: Double = 0.10

    /// Minimum score to qualify as evidence.
    private static let evidenceThreshold: Double = 0.15

    // MARK: - Dependencies

    private let memoryCardDAO: MemoryCardDAO
    private let aiBrain: AIBrainProvider

    static let noMemoryFallback = "I don't remember you telling me that yet."

    init(memoryCardDAO: MemoryCardDAO = MemoryCardDAO(), aiBrain: AIBrainProvider = FakeAIBrain()) {
        self.memoryCardDAO = memoryCardDAO
        self.aiBrain = aiBrain
    }

    // MARK: - Recall

    /// Execute a full recall query. Returns (answer, evidence cards used).
    func recall(question: String) async throws -> RecallResult {
        // 1. Classify query
        let queryClass = classifyQuery(question)

        // 2. If general friend, skip evidence retrieval
        if queryClass == .generalFriend {
            let reply = try await aiBrain.generateFriendReply(transcript: question, recentMemories: [])
            return RecallResult(answer: reply, evidence: [], queryClass: queryClass)
        }

        // 3. Extract query signals
        let signals = extractSignals(from: question)

        // 4. Retrieve candidates (metadata prefilter + text search)
        let candidates = try memoryCardDAO.fetch(filter: MemoryFilter(
            searchText: question,
            limit: 20
        ))

        // 5. Score and rank candidates
        let scored = candidates.map { card in
            ScoredCard(card: card, score: computeScore(card: card, signals: signals))
        }
        .sorted { $0.score > $1.score }
        .filter { $0.score >= Self.evidenceThreshold }

        // 6. Top 3-5 evidence cards
        let topEvidence = Array(scored.prefix(5)).map(\.card)

        // 7. If no evidence, return exact fallback
        if topEvidence.isEmpty {
            return RecallResult(answer: Self.noMemoryFallback, evidence: [], queryClass: queryClass)
        }

        // 8. Generate evidence-bound answer
        let answer = try await aiBrain.answerFromEvidence(question: question, evidence: topEvidence)
        return RecallResult(answer: answer, evidence: topEvidence, queryClass: queryClass)
    }

    // MARK: - Query Classification

    private func classifyQuery(_ question: String) -> QueryClass {
        let lower = question.lowercased()

        // Memory recall keywords
        let recallKeywords = ["what did i", "when did i", "did i say", "did i mention",
                              "what was", "who was", "where was", "how did i",
                              "tell me about", "remind me", "what happened"]

        // Reflection keywords
        let reflectionKeywords = ["pattern", "how often", "what do i usually",
                                  "what's my", "trend", "most common", "frequently"]

        for keyword in recallKeywords {
            if lower.contains(keyword) { return .memoryRecall }
        }
        for keyword in reflectionKeywords {
            if lower.contains(keyword) { return .reflection }
        }

        return .generalFriend
    }

    // MARK: - Signal Extraction

    private struct QuerySignals {
        var keywords: [String]
        var normalizedQuery: String
    }

    private func extractSignals(from question: String) -> QuerySignals {
        let normalized = question.lowercased()
            .replacingOccurrences(of: "[^a-z0-9 ]", with: "", options: .regularExpression)

        let stopWords: Set<String> = ["the", "a", "an", "is", "was", "are", "were", "i", "my", "me",
                                       "did", "do", "what", "when", "where", "who", "how", "about",
                                       "tell", "remind", "say", "said", "think"]

        let keywords = normalized.split(separator: " ")
            .map(String.init)
            .filter { !stopWords.contains($0) && $0.count > 2 }

        return QuerySignals(keywords: keywords, normalizedQuery: normalized)
    }

    // MARK: - Scoring

    private func computeScore(card: MemoryCard, signals: QuerySignals) -> Double {
        let lexical = lexicalScore(card: card, keywords: signals.keywords)
        let tagMatch = tagMatchScore(card: card, keywords: signals.keywords)
        let recency = recencyScore(card: card)
        let importance = Double(card.importanceScore) / 10.0
        let confidence = card.confidenceScore

        return (Self.weightLexical * lexical)
             + (Self.weightTagMatch * tagMatch)
             + (Self.weightRecency * recency)
             + (Self.weightImportance * importance)
             + (Self.weightConfidence * confidence)
    }

    private func lexicalScore(card: MemoryCard, keywords: [String]) -> Double {
        guard !keywords.isEmpty else { return 0 }
        let text = (card.title + " " + card.summary).lowercased()
        let matches = keywords.filter { text.contains($0) }.count
        return Double(matches) / Double(keywords.count)
    }

    private func tagMatchScore(card: MemoryCard, keywords: [String]) -> Double {
        guard !keywords.isEmpty else { return 0 }
        let allTags = (card.emotionTags + card.topicTags + card.peopleTags + card.goalTags)
            .map { $0.lowercased() }
        let matches = keywords.filter { kw in allTags.contains { $0.contains(kw) } }.count
        return Double(matches) / Double(keywords.count)
    }

    private func recencyScore(card: MemoryCard) -> Double {
        let daysSince = Date().timeIntervalSince(card.createdAt) / 86400.0
        // Exponential decay: 1.0 for today, ~0.5 after 7 days, ~0.1 after 30 days
        return max(0, exp(-daysSince / 10.0))
    }
}

// MARK: - Supporting Types

struct ScoredCard {
    let card: MemoryCard
    let score: Double
}

struct RecallResult {
    let answer: String
    let evidence: [MemoryCard]
    let queryClass: RecallEngine.QueryClass
}
